### PR TITLE
[SYCL-MLIR: Uniformity analysis (add inter-procedural tests)

### PR DESCRIPTION
**Pull request by etiotto**
Test inter-procedural dataflow propagation of uniformity.

### CODE REVIEW

**Comment by etiotto** [comment]
This is based on https://github.com/intel/llvm/pull/10141

File: polygeist/include/mlir/Dialect/Polygeist/Analysis/UniformityAnalysis.h
Comment (whitneywhtsang) [inline polygeist/include/mlir/Dialect/Polygeist/Analysis/UniformityAnalysis.h:1]: can you please remove `intra-procedural` at line 9.

Comment (etiotto) [reply polygeist/include/mlir/Dialect/Polygeist/Analysis/UniformityAnalysis.h:1]: done

File: polygeist/test/polygeist-opt/uniformity.mlir [hunk 1]
```diff
@@ -176,3 +176,69 @@ func.func @test2(%cond: i1, %val: i64, %arg1: memref<?x!sycl_nd_item_2>)  {
...
+!sycl_nd_item_2 = !sycl.nd_item<[2], (!sycl_item_2, !sycl_item_2, !sycl_group_2)>
+
+gpu.module @device_func {
+
+// COM: Check the uniformity inter-procedurally.
+func.func private @test3(%cond: i1, %uniform_val: i64, %non_uniform_val : i64)  {
+  %alloca = memref.alloca() : memref<10xi64>
+
+  // COM: The stored value is known inter-procedurally to be uniform.
+  // CHECK: test2_load1, uniformity: uniform
```
Comment (whitneywhtsang) [inline polygeist/test/polygeist-opt/uniformity.mlir:199]:
```suggestion
  // CHECK: test3_load1, uniformity: uniform
```
Also the CHECKs below.

Comment (etiotto) [reply polygeist/test/polygeist-opt/uniformity.mlir]: done I checked the traces and they were all passing but the CHECKs were not updated.

**Review by whitneywhtsang — APPROVED** [review]

File: polygeist/include/mlir/Dialect/Polygeist/Analysis/UniformityAnalysis.h [hunk 2]
```diff
@@ -6,9 +6,9 @@
 //
 //===----------------------------------------------------------------------===//
 //
-// This file contains a simple intra-procedural uniformity analysis which
-// attempts to classify values as either uniform (all threads agree on the
-// content of the value) or not uniform.
+// This file contains an dataflow analysis which attempts to classify values as
```
Comment (sommerlukas) [inline polygeist/include/mlir/Dialect/Polygeist/Analysis/UniformityAnalysis.h:9]: Minor nit.
```suggestion
// This file contains a dataflow analysis which attempts to classify values as
```

**Review by sommerlukas — APPROVED** [review]
Out of curiosity: What did you have to do to enable inter-procedural analysis? Do we get this for "free" from the MLIR dataflow analysis framework?

**Review by jopperm — APPROVED** [review]

**Comment by whitneywhtsang** [comment]
> Out of curiosity: What did you have to do to enable inter-procedural analysis? Do we get this for "free" from the MLIR dataflow analysis framework?

Yes, we get this for "free" from the MLIR dataflow analysis framework.

**Pull request merged**